features.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"github.com/go-errors/errors"
+	"math"
+	"sync/atomic"
+)
+
+// globalFeaturesMap is a map which binds the name of the global feature with it
+// index. The index is just an order of the feature and the final binary
+// representation of feature vector is determined by decode function.
+var globalFeaturesMap = nameToIndex{}
+
+// localFeaturesMap is a map which binds the name of the local feature with it
+// index. The index is just an order of the feature and the final binary
+// representation of feature vector is determined by decode function.
+var localFeaturesMap = nameToIndex{}
+
+// featureFlag represent the status of the feature optional/required and needed
+// to allows future incompatible changes, or backward compatible changes.
+type featureFlag uint8
+
+// featureName represent the name of the feature and needed in order to have the
+// compile errors if we specify wrong feature name.
+type featureName string
+
+type nameToIndex map[featureName]int
+type indexToFlag map[int]featureFlag
+
+const (
+	// optionalFlag represent the feature which we might be used already but
+	// it isn't required yet, and if remote peer doesn't have this feature
+	// we may turn it off without disconnecting with peer.
+	optionalFlag featureFlag = 2 // 0b10
```

Comment:
As far as I see from comment on this [PR](https://github.com/lightningnetwork/lightning-rfc/pull/102) the initial misconception was resolved.